extensions/amp-ad/0.1/amp-ad-3p-impl.js

```diff
@@ -0,0 +1,438 @@
+ */
+const LOADING_ADS_WIN_ID_ = '3pla';
+
+/**
+ * @param {!Element} element
+ * @param {!Window} win
+ * @return {number|boolean}
+ */
+export function allowRenderOutsideViewport(element, win) {
+  // Store in window Object that serves as a set of timers associated with
+  // waiting elements.
+  const loadingAds = win[LOADING_ADS_WIN_ID_] || {};
+  // If another ad is currently loading we only load ads that are currently
+  // in viewport.
+  for (const key in loadingAds) {
+    if (Object.prototype.hasOwnProperty.call(loadingAds, key)) {
+      return false;
+    }
+  }
+
+  // Ad opts into lazier loading strategy where we only load ads that are
+  // at closer than 1.25 viewports away.
+  if (element.getAttribute('data-loading-strategy') ==
+      'prefer-viewability-over-views') {
+    return 1.25;
+  }
+  return true;
+}
+
+/**
+ * Decrements loading ads count used for throttling.
+ * @param {number} timerId of timer returned from incrementLoadingAds
+ * @param {!Window} win
+ */
+export function decrementLoadingAds(timerId, win) {
+  timer.cancel(timerId);
+  const loadingAds = win[LOADING_ADS_WIN_ID_];
+  if (loadingAds) {
+    delete loadingAds[timerId];
+  }
+}
+
+/**
+ * Increments loading ads count for throttling.
+ * @param {!Window} win
+ * @return {number} timer ID for testing
+ */
+export function incrementLoadingAds(win) {
+  let loadingAds = win[LOADING_ADS_WIN_ID_];
+  if (!loadingAds) {
+    loadingAds = {};
+    win[LOADING_ADS_WIN_ID_] = loadingAds;
+  }
+
+  const timerId = timer.delay(() => {
+    // Unfortunately we don't really have a good way to measure how long it
+    // takes to load an ad, so we'll just pretend it takes 1 second for
+    // now.
+    decrementLoadingAds(timerId, win);
+  }, 1000);
+  loadingAds[timerId] = 1;
+  return timerId;
+}
+
+/**
+ * @param {!Element} el
+ * @param {!Window} win
+ * @return {boolean} whether element or its ancestors have position
+ * fixed (unless they are POSITION_FIXED_TAG_WHITELIST).
+ * This should only be called when a layout on the page was just forced
+ * anyway.
+ */
+export function isPositionFixed(el, win) {
+  let hasFixedAncestor = false;
+  do {
+    if (POSITION_FIXED_TAG_WHITELIST[el.tagName]) {
+      return false;
+    }
+    if (win/*because only called from onLayoutMeasure */
+            ./*OK*/getComputedStyle(el).position == 'fixed') {
+      // Because certain blessed elements may contain a position fixed
+      // container (which contain an ad), we continue to search the
+      // ancestry tree.
+      hasFixedAncestor = true;
+    }
+    el = el.parentNode;
+  } while (el && el.tagName != 'BODY');
+  return hasFixedAncestor;
+}
+
+/** @const {!string} Tag name for 3P AD implementation. */
+export const TAG_3P_IMPL = 'amp-ad-3p-impl';
+
+export class AmpAd3PImpl extends AMP.BaseElement {
+
+  /** @override */
+  getPriority() {
+    // Loads ads after other content.
+    return 2;
+  }
+
+  renderOutsideViewport() {
```

Comment:
Ha, looks like I commented on the wrong line. You got it though, I meant `#isReadyToBuild`.